packages/graphql/src/services/practiceQuizzes.ts

```diff
@@ -27,32 +39,247 @@ export async function getPracticeQuizData(
+    return null
+  }
+
+  // TODO: also handle the case where the participant is logged in, but has no valid participation in the considered course
+  if (!ctx.user?.sub) {
+    // TODO: if user is not logged in, do not check the questionResponse, only save it on the instance / update the instance results
+    return null
+  }
+
+  // update the aggregated data on the element instance
+  await ctx.prisma.elementInstance.update({
+    where: {
+      id,
+    },
+    data: {
+      results: {
+        wrong: existingInstance.results.wrong + (correctness === 0 ? 1 : 0),
+        partial: existingInstance.results.partial + (correctness === 1 ? 1 : 0),
+        correct: existingInstance.results.correct + (correctness === 2 ? 1 : 0),
+        total: existingInstance.results.total + 1,
+      },
+    },
+  })
+
+  // find existing question response to this instance by this user
+  const existingResponse = await ctx.prisma.questionResponse.findUnique({
+    where: {
+      participantId_elementInstanceId: {
+        participantId: ctx.user.sub,
+        elementInstanceId: id,
+      },
+    },
+  })
+
+  ctx.prisma.questionResponseDetail.create({
+    data: {
+      response: {
+        correctness: correctness,
+      },
+      participant: {
+        connect: { id: ctx.user.sub },
+      },
+      questionInstance: {
+        connect: { id },
+      },
+      participation: {
+        connect: {
+          courseId_participantId: {
+            courseId,
+            participantId: ctx.user.sub,
+          },
+        },
+      },
+    },
+  })
+
+  // TODO: think about possibility to fuse these two cases with a single upsert (however, keep in mind that we need to update JSON stuff here)
+  if (existingResponse) {
+    let aggregatedResponse =
+      existingResponse.aggregatedResponses as AggregatedResponseFlashcard
+    switch (correctness) {
+      case 0:
+        aggregatedResponse = {
+          ...aggregatedResponse,
+          wrong: aggregatedResponse.wrong + 1,
+          total: aggregatedResponse.total + 1,
+        }
+        break
+      case 1:
+        aggregatedResponse = {
+          ...aggregatedResponse,
+          partial: aggregatedResponse.partial + 1,
+          total: aggregatedResponse.total + 1,
+        }
+        break
+      case 2:
+        aggregatedResponse = {
+          ...aggregatedResponse,
+          correct: aggregatedResponse.correct + 1,
+          total: aggregatedResponse.total + 1,
+        }
+        break
+    }
+
+    // update existing response
+    const questionResponse = await ctx.prisma.questionResponse.update({
+      where: {
+        participantId_elementInstanceId: {
+          participantId: ctx.user.sub,
+          elementInstanceId: id,
+        },
+      },
+      data: {
+        trialsCount: {
+          increment: 1,
+        },
+        response: {
+          correctness,
+        },
+        aggregatedResponses: aggregatedResponse,
+        correctCount: {
```

Comment:
would we need partial correct count as well, or is having only correct count sufficient?

Reply:
will probably depend heavily on the type of repetition algorithm we use, but to have full flexibility, it would probably make sense to track this number as well